lib/miq_automation_engine/service_models/miq_ae_service_container_deployment.rb

```diff
@@ -0,0 +1,172 @@
+    expose :container_deployment_nodes, :association => true
+    expose :deployed_ems, :association => true
+    expose :deployed_on_ems, :association => true
+    expose :automation_task, :association => true
+    expose :roles_addresses
+    expose :container_nodes_by_role
+    expose :ssh_auth
+    expose :rhsm_auth
+
+    def assign_container_deployment_node(vm_id, role)
+      container_nodes_by_role(role).each do |deployment_node|
+        next unless deployment_node.vm_id.nil?
+        deployment_node.add_vm vm_id
+      end
+    end
+
+    def rhsm_creds
+      [rhsm_auth.userid, rhsm_auth.password]
+    end
+
+    def ssh_user
+      ssh_auth.userid
+    end
+
+    def add_deployment_provider(options)
+      object_send(:add_deployment_provider, options)
+    end
+
+    def regenerate_ansible_inventory
+      object_send(:generate_ansible_yaml)
+    end
+
+    def regenerate_ansible_subscription_inventory
+      object_send(:generate_ansible_inventory_for_subscription)
+    end
+
+    def add_automation_task(task)
+      ar_method do
+        wrap_results(@object.automation_task = AutomationTask.find_by_id(task.id))
+        @object.save!
+      end
+    end
+
+    def customize(options = {})
+      ar_method do
+        if @object.customizations.empty?
+          @object.customizations = {:agent => {}}
+        end
+        options.each do |key, val|
+          wrap_results(@object.customizations[:agent][key] = val)
+        end
+        @object.save!
+      end
+    end
+
+    def perform_scp(ip, username, local_path, remote_path)
+      Net::SCP.upload!(ip,
+                       username,
+                       local_path,
+                       remote_path,
+                       :ssh => {:key_data => ssh_auth.auth_key})
+    end
+
+    def perform_agent_commands(ip, username, commands = [])
+      pp commands
+      ssh = LinuxAdmin::SSH.new(ip, username, ssh_auth.auth_key)
+      result = ssh.perform_commands(commands)
+      result
+    end
+
+    def check_connection(ip, username, ips)
+      agent = LinuxAdmin::SSHAgent.new(ssh_auth.auth_key, "/tmp/ssh_manageiq/ssh_manageiq_#{id}")
+      success, unreachable_ips = agent.with_service do |socket|
+        ssh = LinuxAdmin::SSH.new(ip, username)
+        unreachable_ips = []
+        conneted = true
+        ips.each do |sub_ip|
+          res = ssh.perform_commands(["sudo -E ssh -o 'StrictHostKeyChecking no' #{username}@#{sub_ip} echo $?"], socket)[:stdout]
+          unless res.include?("0\r\n")
+            conneted = false
+            unreachable_ips << sub_ip
+          end
+        end
+        [conneted, unreachable_ips]
+      end
+      unless success
+        raise StandardError, "couldn't connect to : #{unreachable_ips.join(',')}"
+      end
+      success
+    end
+
+    def playbook_running?
+      pid = ar_method do
+        @object.customizations[:agent][:agent_pid] if @object.customizations[:agent]
+      end
```

Comment:
Also `playbook_running?` may be misleading name.
`@object.customizations` is remembered since last `run_playbook_command`, right?
How about `ansible_exists?` or better `playbook_was_started?`